zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,156 @@
+ * Authorization is done by checking with clients' URI (uniform resource identifier) in ACL metadata for matched domains.
+ * This class is meant to support the following use patterns:
+ *  1.Multiple creators, one reader - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read by an identity from a different domain.
+ *  2.Multiple creators, multiple readers - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read by multiple identities from multiple domains.
+ *  3.Multiple creators, one all-permission accessor - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read/update/deleted by an identity from a different domain.
+ * Optional features include:
+ *    set client Id as znode ACL when creating nodes,
+ *    add additional public read ACL to znodes based on path.
+ */
+public class X509ZNodeGroupAclProvider extends ServerAuthenticationProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(X509ZNodeGroupAclProvider.class);
+  private final String logStrPrefix = this.getClass().getName() + ":: ";
+  private final X509TrustManager trustManager;
+  private final X509KeyManager keyManager;
+  static final String ZOOKEEPER_ZNODEGROUPACL_SUPERUSER = "zookeeper.znodeGroupAcl.superUser";
+
+  public X509ZNodeGroupAclProvider() {
+    ZKConfig config = new ZKConfig();
+    this.keyManager = X509AuthenticationUtil.createKeyManager(config);
+    this.trustManager = X509AuthenticationUtil.createTrustManager(config);
+  }
+
+  public X509ZNodeGroupAclProvider(X509TrustManager trustManager, X509KeyManager keyManager) {
+    this.trustManager = trustManager;
+    this.keyManager = keyManager;
+  }
+
+  @Override
+  public KeeperException.Code handleAuthentication(ServerObjs serverObjs, byte[] authData) {
+    ServerCnxn cnxn = serverObjs.getCnxn();
+    X509Certificate clientCert;
+    try {
+      clientCert = X509AuthenticationUtil.getAndAuthenticateClientCert(cnxn, trustManager);
+    } catch (KeeperException.AuthFailedException e) {
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    // Extract URI from certificate
+    String uri;
+    try {
+      uri = X509AuthenticationUtil.getClientId(clientCert);
+    } catch (Exception e) {
+      // Failed to extract URI from certificate
+      LOG.error(logStrPrefix + "Failed to extract URI from certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), e);
+      return KeeperException.Code.OK;
+    }
+
+    // User belongs to super user group
+    if (uri.equals(System.getProperty(ZOOKEEPER_ZNODEGROUPACL_SUPERUSER))) {
+      cnxn.addAuthInfo(new Id("super", uri));
+      LOG.info("Authenticated Id '{}' as super user", uri);
+      return KeeperException.Code.OK;
+    }
+
+    // Get authorized domain names for client
+    ClientUriDomainMappingHelper uriDomainMappingHelper =
+        new ZkClientUriDomainMappingHelper(serverObjs.getZks());
```

Comment:
Is there a reason we're recreating this object every time `handleAuth()` is called?